vim-patch:8.0.{1356,1491,1495,1522,1538,1540},8.1.{554,670,1300,1303,1875,2377}

janlazo opened the pull request:
8.0.1491 is insufficent by itself but I had issues with the tests when merging the followup popupmnu.c patches.

I don't know what do with the other 8.0.x patches. I'm going straight  to 8.1.x patches.

janlazo committed:
vim-patch:8.0.1491: the minimum width of the popup menu is hard coded

Problem:    The minimum width of the popup menu is hard coded.
Solution:   Add the 'pumwidth' option. (Christian Brabandt, James McCoy,
            closes vim/vim#2314)
https://github.com/vim/vim/commit/a8f04aa275984183bab5bb583b128f38c64abb69

janlazo committed:
vim-patch:8.0.1495: having 'pumwidth' default to zero has no merit

Problem:    Having 'pumwidth' default to zero has no merit.
Solution:   Make the default 15, as the actual default value.
https://github.com/vim/vim/commit/42443c7d7fecc3a2a72154bb6139b028438617c2

Includes 'pumwidth' documentation changes from 8.0.1531.
Sort 'pum*' option in alphabetical order.

janlazo committed:
vim-patch:8.1.0670: macro for popup menu width is unused

Problem:    Macro for popup menu width is unused.
Solution:   Remove it. (Hirohito Higashi)
https://github.com/vim/vim/commit/3d631cb0b34b03c7bdf45ad852d3644c7cf62743

janlazo committed:
vim-patch:8.0.1522: popup menu is positioned in the wrong place

Problem:    Popup menu is positioned in the wrong place. (Davit Samvelyan,
            Boris Staletic)
Solution:   Correct computation of the column and the conditions for that.
            (Hirohito Higashi, closes vim/vim#2640)
https://github.com/vim/vim/commit/4287ed33ddc324d26dd05d3e19596dd74cf479d6

janlazo committed:
vim-patch:8.0.1538: popupmenu is too far left when completion is long

Problem:    Popupmenu is too far left when completion is long. (Linwei)
Solution:   Adjust column computations. (Hirohito Higashi, closes vim/vim#2661)
https://github.com/vim/vim/commit/bb008dd3239c5fe3ac04501e38e4c950fa9426c8

janlazo commented on runtime/autoload/provider/clipboard.vim:
```diff
@@ -172,6 +172,10 @@ function! s:clipboard.set(lines, regtype, reg) abort
   if jobid > 0
     call jobsend(jobid, a:lines)
     call jobclose(jobid, 'stdin')
+    if argv[0] ==# 'xclip'
```
Tested this with by copying some text via Vim's clipboard first and then running `make oldtest`.

justinmk commented on src/nvim/screen.c:
```diff
@@ -121,7 +121,6 @@
 #define MB_FILLER_CHAR '<'  /* character used when a double-width character
                              * doesn't fit. */
-#define W_ENDCOL(wp)   (wp->w_wincol + wp->w_width)
 #define W_ENDROW(wp)   (wp->w_winrow + wp->w_height)
```
i would perhaps move `W_ENDROW` along with `W_ENDCOL`..

justinmk commented on test/functional/ui/popupmenu_spec.lua:
```diff
@@ -1204,10 +1204,10 @@ describe('builtin popupmenu', function()
+      {1:^~                    }{n: word      }|
+      {1:~                    }{n: choice    }|
+      {1:~                    }{s: text      }|
+      {1:~                    }{n: thing     }|
```
why did this change? Did `PUM_DEF_WIDTH` not function in the same manner as 'pumwidth' now does?

Perhaps it's worth adding an extra test with a non-default 'pumwidth'?

janlazo replied:
These change again in later popupmnu.c 8.0.x patches. vim-patch:8.0.1491 added new code to handle non-default 'pumwidth',

janlazo replied:
complete() `startcol`  + `&pumwidth` must be less than the screen width to display the pum at the `startcol`.

janlazo committed:
vim-patch:8.1.0554: popup menu overlaps with preview window

Problem:    Popup menu overlaps with preview window.
Solution:   Adjust the height computation. (Hirohito Higashi, closes vim/vim#3414)
https://github.com/vim/vim/commit/614ab8aa00346724bfc27980d25985d482269b75

Cherry-picked "row -> pum_win_row" rename changes from patch 8.1.0062.

janlazo committed:
screen: fix pvs/v1048

janlazo committed:
vim-patch:8.1.1303: not possible to hide a balloon

Problem:    Not possible to hide a balloon.
Solution:   Hide the balloon when balloon_show() is called with an empty
            string or list.  Add balloon_gettext().
https://github.com/vim/vim/commit/be0a2597ae0d9eb0b8a8a2fc9ae1784faa929844

janlazo committed:
fixup! vim-patch.sh: list related missing Vim patches [ci skip] #11514

janlazo committed:
vim-patch:8.0.1540: popup menu positioning fails with longer string

Problem:    Popup menu positioning fails with longer string.
Solution:   Only align with right side of window when width is less than
            'pumwidth' (closes vim/vim#2661)
https://github.com/vim/vim/commit/2b10bcbfc1c025bf7e6358326ee70105e7d30e96

justinmk commented on runtime/doc/options.txt:
```diff
@@ -4529,6 +4529,13 @@ A jump table for the options with a short description can be found at |Q_op|.
+	|ins-completion-menu|.
+
+						*'pumwidth'* *'pw'*
+'pumwidth' 'pw'		number	(default 0)
```
technically this should go after 'pumblend', these entries are alphabetical.

janlazo committed:
vim-patch:8.1.1875: cannot get size and position of the popup menu

Problem:    Cannot get size and position of the popup menu.
Solution:   Add pum_getpos(). (Ben Jackson, closes vim/vim#4827)
https://github.com/vim/vim/commit/e9bd57286a5cbb0e1ec18b5d194dc4af1bda9f3a

https://github.com/neovim/neovim/pull/11562 backported the vim patch.
This patch only updates the runtime/doc/ files to match Vim.

justinmk commented on test/functional/ui/popupmenu_spec.lua:
```diff
@@ -1156,10 +1156,10 @@ describe('builtin popupmenu', function()
+      {n:word           }{1:                 }|
+      {n:choice         }{1:                 }|
+      {n:text           }{1:                 }|
+      {n:thing          }{1:                 }|
```
why did the menu position move to the left?

janlazo replied:
https://github.com/neovim/neovim/pull/11616#discussion_r361896362

justinmk commented:
`src/nvim/testdir/test_popup.vim` doesn't test anything for us since it's using Vim-only `term_xx` calls. Would be worth adding another screen test to cover non-defualt 'pumwidth'.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -3770,6 +3771,8 @@ feedkeys({string} [, {mode}])				*feedkeys()*
 		feedkeys("\<CR>") simulates pressing of the <Enter> key. But
 		feedkeys('\<CR>') pushes 5 characters.
+		A special code that might be useful is <Ignore>, it exits the
+		wait for a character without doing anything.  *<Ignore>*
```
Do we actually support this? Is there a test for it?

janlazo replied:
It's for `K_IGNORE`. `git grep --files-with-matches '<Ignore>'` returns `runtime matches` so it's probably untested.

janlazo replied:
I missed [8.1.1300](https://github.com/vim/vim/commit/2f10658b06bbdd8f25c4ff152266c808234cee0a)

janlazo committed:
ui: add basic tests for pumheight,pumwidth

janlazo committed:
vim-patch:8.1.1300: in a terminal 'ballooneval' does not work right away

Problem:    In a terminal 'ballooneval' does not work right away.
Solution:   Flush output after drawing the balloon.  Add the <Ignore> key
            code.  Add a test.
https://github.com/vim/vim/commit/2f10658b06bbdd8f25c4ff152266c808234cee0a

janlazo committed:
vim-patch:8.1.2377: GUI: when losing focus a pending operator is executed

Problem:    GUI: when losing focus a pending operator is executed.
Solution:   Do not execute an operator when getting K_IGNORE. (closes vim/vim#5300)
https://github.com/vim/vim/commit/fa5612c7d836eb789e0f8ff4b10461b8640a14b2

janlazo committed:
vim-patch:8.0.1356: using simalt in a GUIEnter autocommand inserts characters

Problem:    Using simalt in a GUIEnter autocommand inserts strange characters.
            (Chih-Long Chang)
Solution:   Ignore K_NOP in Insert mode. (closes vim/vim#2379)
https://github.com/vim/vim/commit/c5aa55db7e5bc791f99fb15b0f4be0d5dd166f62

justinmk commented on runtime/doc/options.txt:
```diff
@@ -4547,6 +4540,19 @@ A jump table for the options with a short description can be found at |Q_op|.
+'pumwidth' 'pw'		number	(default 15)
+			global
+	Determines the minium width to use for the popup menu for Insert mode
+	completion.  |ins-completion-menu|.
```
would be great to also mention your comment in the doc here:

> complete() `startcol`  + `&pumwidth` must be less than the screen width to display the pum at the `startcol`.

something like:

> If the cursor column  + 'pumwidth' exceed screen width, the popup menu is nudged to fit on the screen.

justinmk replied:
done in https://github.com/neovim/neovim/pull/11604

justinmk commented on src/nvim/keymap.c:
```diff
@@ -309,6 +309,7 @@ static const struct key_name_entry {
   { K_ZERO,            "Nul" },
   { K_SNR,             "SNR" },
   { K_PLUG,            "Plug" },
+  { K_IGNORE,          "Ignore" },
```
Ah, `K_IGNORE` already existed, but now it's exposed as `<Ignore>` keycode.
